common/cuda_hip/base/kernel_launch_reduction.hpp.inc

```diff
@@ -0,0 +1,534 @@
+                         op);
+        if (threadIdx.x == 0) {
+            storage[blockIdx.x] = finalize(partial);
+        }
+    }
+}
+
+
+template <typename ValueType, typename KernelFunction, typename ReductionOp,
+          typename FinalizeOp, typename... KernelArgs>
+__global__ __launch_bounds__(
+    default_block_size) void generic_kernel_reduction_2d(int64 rows, int64 cols,
+                                                         KernelFunction fn,
+                                                         ReductionOp op,
+                                                         FinalizeOp finalize,
+                                                         ValueType identity,
+                                                         ValueType* storage,
+                                                         KernelArgs... args)
+{
+    __shared__
+        UninitializedArray<ValueType, default_block_size / config::warp_size>
+            warp_partial;
+    static_assert(default_block_size / config::warp_size <= config::warp_size,
+                  "needs third reduction level");
+    auto tidx = thread::get_thread_id_flat<int64>();
+    auto grid_size = thread::get_thread_num_flat<int64>();
+    auto warp =
+        group::tiled_partition<config::warp_size>(group::this_thread_block());
+    auto partial = identity;
+    for (int64 i = tidx; i < rows * cols; i += grid_size) {
+        const auto row = i / cols;
+        const auto col = i % cols;
+        partial = op(partial, fn(row, col, args...));
+    }
+    partial = reduce(warp, partial, op);
+    if (warp.thread_rank() == 0) {
+        warp_partial[threadIdx.x / config::warp_size] = partial;
+    }
+    __syncthreads();
+    if (threadIdx.x < config::warp_size) {
+        partial = reduce(warp,
+                         threadIdx.x < default_block_size / config::warp_size
+                             ? warp_partial[threadIdx.x]
+                             : identity,
+                         op);
+        if (threadIdx.x == 0) {
+            storage[blockIdx.x] = finalize(partial);
+        }
+    }
+}
+
+
+template <typename ValueType, typename KernelFunction, typename ReductionOp,
+          typename FinalizeOp, typename... KernelArgs>
+void run_kernel_reduction_cached(std::shared_ptr<const DefaultExecutor> exec,
+                                 KernelFunction fn, ReductionOp op,
+                                 FinalizeOp finalize, ValueType identity,
+                                 ValueType* result, size_type size,
+                                 Array<char>& tmp, KernelArgs&&... args)
+{
+    constexpr int oversubscription = 16;
+    constexpr auto block_size = default_block_size;
+    const auto num_blocks = std::min<int64>(
+        ceildiv(size, block_size), exec->get_num_warps() * oversubscription);
+    if (num_blocks > 1) {
+        const auto required_storage = sizeof(ValueType) * num_blocks;
+        tmp.set_executor(exec);
```

Comment:
I could do this either on the `core` side or the kernel side. If users pass in an Array on the wrong executor, I need to somehow turn it into a correct one. Either by checking on the host side and creating a different array if necessary, or by changing the executor of the input array. Also by set_executor we might copy over data, so `tmp.clear(); tmp.set_executor(exec);` might be more appropriate?

Reply:
I mean this kind of transparently transfers the data to the executor, with allocations, which the user probably did not intend to do, if they actually used the advanced version of the `_impl` function where they explicitly passed in a temp array. So, I am suggesting we throw instead of continue if there is an executor mismatch.

Reply:
I think this is probably not the right point to change the executor, so I'll move the operation to `core`. If `tmp` was initialized without executor, we would still need to add some special-case treatment code. How about `if (exec != array.get_executor()) { array.clear(); array.set_executor(exec); }`?